ls1intum/Artemis
# Feature/Add new automatic submission run property to programming exercises

### Checklist

### Motivation and Context



For https://github.com/ls1intum/Artemis/issues/770: Trigger a submission run when the due date of a programming exercise expires.
This PR provides an updated model for storing the submission run date and integrates the configuration in the client. The LifecycleHook itself will be integrated in another PR.

### Description


Introduces the new property automatic_submission_run_date, but not directly in the exercise table. Instead the new entity programming_exercise_details table is linked to the entity ProgrammingExercise with the `@SecondaryTable` annotation. This feature could be used in future for more properties that are exercise type specific.

The new property is a date, that can be set freely when the due date is set (default date = due date)

### Steps for Testing


1. Log in to Artemis
2. Create a new Programming Exercise / Link new Programming Exercise / View Programming Exercise / Edit Programming Exercise
3. Enable/Disable the new property and check if it is reflected in the network calls / database

### Screenshots



**Option is inactive:** 

![image](https://user-images.githubusercontent.com/28230611/64422516-8ec50000-d0a4-11e9-9c33-8d2af2a4f110.png)

**Option is active:**

![image](https://user-images.githubusercontent.com/28230611/64628463-bcd27900-d3f1-11e9-9cb5-ee0be88c8823.png)

**Error scenario: DueDate after buildDate**

![image](https://user-images.githubusercontent.com/28230611/64689786-556b0680-d48f-11e9-91e6-dc57086e5757.png)
